File: MCTS_BlokusDuo_cpp/Move.cpp
//
// Created by joaovieira on 12/24/19.
//

#include <sstream>
#include "Move.h"

Move::Move(std::shared_ptr<GamePiece> gamePiece, Position *center) {
    this->gamePiece = gamePiece;
    this->center = center;
}

Move::~Move() {
    delete center;
}

std::shared_ptr<GamePiece> Move::getGamePiece() {
    return gamePiece;
}

Position *Move::getCenter() {
    return center;
}

std::string Move::moveToString(Move *move) {
    std::stringstream text;

    if (move == nullptr) {
        text << "0000";
        return text.str();
    }

    text << std::hex << move->center->getX() + 1;
    text << std::hex << move->center->getY() + 1;
    text << move->gamePiece->getCodeName();
    text << move->gamePiece->getTransformation();

    return text.str();
}

bool Move::equals(Move *move) {
    if (this == move) return true;
    if (move == nullptr) return false;
    return move->center->equals(center) && move->gamePiece->equals(gamePiece.get());
}

Move::Move(Move *move) {
    gamePiece = move->gamePiece;
    center = new Position(move->center);
}

File: MCTS_BlokusDuo_java/src/com/joaomiguelvieira/Move.java
package com.joaomiguelvieira;

import java.util.Objects;

public class Move {
    private GamePiece gamePiece;
    private Position center;

    public Move(GamePiece gamePiece, Position center) {
        this.gamePiece = gamePiece;
        this.center = new Position(center);
    }

    public GamePiece getGamePiece() {
        return gamePiece;
    }

    public Position getCenter() {
        return center;
    }

    public static String moveToString(Move move) {
        if (move == null)
            return "0000";

        String string = "";
        string += Integer.toHexString(move.center.getX() + 1);
        string += Integer.toHexString(move.center.getY() + 1);
        string += move.gamePiece.getCodeName();
        string += move.gamePiece.getTransformation();

        return string;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Move move = (Move) o;
        return gamePiece.equals(move.gamePiece) &&
                center.equals(move.center);
    }

    @Override
    public int hashCode() {
        return Objects.hash(gamePiece, center);
    }
}
